lib/skate/detours/missed_stops.ex

```diff
@@ -92,99 +93,63 @@ defmodule Skate.Detours.MissedStops do
           shape :: [Util.Location.From.t()],
           stops :: [Util.Location.From.t()]
         ) :: [Skate.Detours.ShapeSegment.t()]
-  defp segment_shape_by_stops([] = _shape, [] = _stops),
+  def segment_shape_by_stops([] = _shape, [] = _stops),
     do: []
 
-  defp segment_shape_by_stops(shape, [] = _stops),
+  def segment_shape_by_stops(shape, [] = _stops),
     # If there are no stops, return the shape
     do: [%Skate.Detours.ShapeSegment{points: shape, stop: :none}]
 
-  defp segment_shape_by_stops(shape, stops) do
-    # Find the stop closest to the shape
-    %{index: anchor_stop_index, shape_dist: %{index: shape_point_anchor_index}} =
-      find_stop_closest_to_shape(shape, stops)
+  def segment_shape_by_stops(shape, stops) do
+    %{stop_index: anchor_stop_index, anchor_point: anchor_point, shape_index: anchor_shape_index} =
+      stops
+      |> Enum.with_index(fn stop, stop_index ->
+        {anchor_point, shape_index} = NearestPoint.nearest_point_on_shape(shape, stop)
+
+        %{
+          stop: stop,
+          stop_index: stop_index,
+          anchor_point: anchor_point,
+          shape_index: shape_index,
+          dist: Location.distance(anchor_point, stop)
+        }
+      end)
+      |> Enum.min_by(fn %{dist: dist} -> dist end)
 
-    # Split the shape and stops at the anchor point
-    # Add one to the Shape Point Anchor Index so that the point ends up in `left_shape`
-    {left_shape, right_shape} = Enum.split(shape, shape_point_anchor_index + 1)
+    {stops_before_anchor, [anchor_stop | stops_after_anchor]} =
+      Enum.split(stops, anchor_stop_index)
 
-    {left_stops, [anchor | right_stops]} = Enum.split(stops, anchor_stop_index)
+    {points_before_anchor, points_after_anchor} = Enum.split(shape, anchor_shape_index + 1)
 
-    # Process `left_shape` into segments
-    {left_segments, [anchor_segment]} =
-      Enum.split(segment_shape_by_stops(left_shape, left_stops), -1)
+    shape_before_anchor = points_before_anchor ++ [anchor_point]
+    shape_after_anchor = [anchor_point] ++ points_after_anchor
 
-    # Take last segment, which should not have a stop,
-    # and reconnect it with the anchor stop
-    anchor_segment = %Skate.Detours.ShapeSegment{
-      (%Skate.Detours.ShapeSegment{stop: :none} = anchor_segment)
-      | stop: anchor
-    }
+    {segments_before_anchor, [incomplete_anchor_segment]} =
+      shape_before_anchor |> segment_shape_by_stops(stops_before_anchor) |> Enum.split(-1)
 
-    left_segments ++ [anchor_segment] ++ segment_shape_by_stops(right_shape, right_stops)
-  end
+    anchor_segment = %Skate.Detours.ShapeSegment{incomplete_anchor_segment | stop: anchor_stop}
```

Comment:
question: I recall that the only reason we noticed that there was an issue originally was because we had the `%Skate.Detours.ShapeSegment{stop: :none} = anchor_segment` assertion, do you think it's worth dropping that assertion? 

I'd rather have it than not personally.

Reply:
Oooohhhoohooo THAT's why we had that assertion!

Yes - let's add that back in

Reply:
[How's this?](https://github.com/mbta/skate/pull/2640/commits/6970a232b0c2a015f534641d36e0cb28cedcd18c)